### PR TITLE
Add end-to-end tests to our CI

### LINKED ISSUE #4234: Implement end to end tests with real server
Implement end to end tests with real server.
~~Exact test perimeter to be determined.~~

This issue aims to provide a minimum test setup, so that upcoming features/bug fixes can implement further tests.
Currently Linux only, then port it to Windows (might be a bit more work due to docker) and Mac (no CI right now).

- [x] set up test server within CI
  - Android example https://github.com/nextcloud/android/blob/master/.drone.yml#L30 server service, available with http://server
- [x] connect client with server
- [ ] Tests
  - [x] create some files, upload them, check that they are on server
  - [x] create some files, upload them, delete them, check that they were deleted on server
  - [ ] create a file, upload it, check activity endpoint, that info is available

### PR DESCRIPTION
**Pull request by claucambra**
Fix #4234

This PR adds end-to-end testing to our CI.

Specifically, it:

1. Adds and sets up a real server to our Drone pipeline
2. Adds a end to end testing utilities class that makes it easier to test with the server set up in the CI
3. Adds a test checking the correct set-up of an account
4. Adds a test checking file transfer (down-syncing an existing file, creating and uploading a new file, and deleting said created file)

Addresses #4234

### CODE REVIEW

File: test/endtoendtestutils.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        QByteArray credHash = QByteArray(_cred->user().toUtf8() + ":" + _cred->password().toUtf8()).toBase64();
+        req.setRawHeader("Authorization", "Basic " + credHash);
+
+        return OCC::AccessManager::createRequest(op, req, outgoingData);
+    }
+
+private:
+    // The credentials object dies along with the account, while the QNAM might
+    // outlive both.
+    QPointer<const EndToEndTestCredentials> _cred;
```
Comment (mgallien) [inline test/endtoendtestutils.cpp:56]: not sure why you use a `QPointer` because you never test its validity
keep in mind that `QPointer` will be set to `nullptr` once the pointed object (if derived from a `QObject`) has been deleted

File: test/teste2efiletransfer.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+class E2eFileTransferTest : public QObject
+{
+    Q_OBJECT
+
+public:
+    E2eFileTransferTest() = default;
+
+private:
+    EndToEndTestHelper _helper;
+    OCC::Folder *_testFolder;
```
Comment (mgallien) [inline test/teste2efiletransfer.cpp:34]: please set an initial value

File: test/endtoendtestutils.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+OCC::Folder *EndToEndTestHelper::configureSyncFolder()
+{
+    if(_syncFolder) {
+        removeConfiguredSyncFolder();
+    }
+
+    qDebug() << "Creating temp end-to-end test folder.";
+    Q_ASSERT(_tempDir.isValid());
+    OCC::FileSystem::setFolderMinimumPermissions(_tempDir.path());
+    qDebug() << "Created temp end-to-end test folder at:" << _tempDir.path();
```
Comment (mgallien) [inline test/endtoendtestutils.cpp:132]: please create a logging category and make use of it
that will allow filtering and easy retrieving during analysis

File: test/endtoendtestutils.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+    setupFolderMan();
+
+    OCC::FolderDefinition definition;
+    definition.localPath = _tempDir.path();
+    definition.targetPath = '/';
```
Comment (mgallien) [inline test/endtoendtestutils.cpp:138]: you will probably want to make this configurable in the tests to allow reproducing cases of bugs with the target path not being the root folder

File: test/teste2efiletransfer.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    void testSyncFolder()
+    {
+        // Try the down-sync first
+        QSignalSpy folderSyncFinished(_testFolder, &OCC::Folder::syncFinished);
+        OCC::FolderMan::instance()->forceSyncForFolder(_testFolder);
+        QVERIFY(folderSyncFinished.wait(3000));
+
+        const auto testFolderPath = _testFolder->path();
+        const QString expectedFilePath(testFolderPath + QStringLiteral("welcome.txt"));
```
Comment (mgallien) [inline test/teste2efiletransfer.cpp:59]: where does this file come from ?

File: test/teste2efiletransfer.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        QVERIFY(OCC::Utility::writeRandomFile(localFilePath));
+
+        OCC::FolderMan::instance()->forceSyncForFolder(_testFolder);
+        QVERIFY(folderSyncFinished.wait(3000));
+        qDebug() << "First folder sync complete";
+
+        const auto waitForServerToProcessTime = QTime::currentTime().addSecs(3);
+        while (QTime::currentTime() < waitForServerToProcessTime) {
+            QCoreApplication::processEvents(QEventLoop::AllEvents, 100);
+        }
```
Comment (mgallien) [inline test/teste2efiletransfer.cpp:76]: why is this needed ?

File: .drone.yml [hunk 7]
```diff
@@ -9,7 +9,8 @@ steps:
       path: /drone/build
   commands:
     - cd /drone/build
-    - cmake -G Ninja -DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11 -DCMAKE_BUILD_TYPE=Debug -DQUICK_COMPILER=ON -DBUILD_UPDATER=ON -DBUILD_TESTING=1 -DECM_ENABLE_SANITIZERS=address -DCMAKE_CXX_FLAGS=-Werror -DOPENSSL_ROOT_DIR=/usr/local/lib64 ../src
+    - cmake -G Ninja -DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11 -DCMAKE_BUILD_TYPE=Debug -DQUICK_COMPILER=ON -DBUILD_UPDATER=ON -DBUILD_TESTING=1 -DBUILD_E2E_TESTS=ON -DECM_ENABLE_SANITIZERS=address -DCMAKE_CXX_FLAGS=-Werror -DOPENSSL_ROOT_DIR=/usr/local/lib64 ../src
```
[the remark this answers is not in the document]
Comment (claucambra) [reply .drone.yml:12]: At the moment we cannot set up the server (or at least, not easily from what I've seen) in our Github actions so building the tests will just slow down our other CI runs -- are we sure we want to do this?

File: test/teste2efiletransfer.cpp (hunk 5 above)
Comment (claucambra) [reply test/teste2efiletransfer.cpp:59]: It ships with the server docker image

File: test/teste2efiletransfer.cpp (hunk 6 above)
Comment (claucambra) [reply test/teste2efiletransfer.cpp:76]: The server will often not be able to process new files immediately after the upload/delete is done, so we sleep for three seconds to give the server a chance to finish what it's doing with the recent client's action, and then we check the action we wanted was completed

File: .drone.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgallien) [reply .drone.yml:12]: but we want style checking or clang-tidy checking on as many source file as we can
so I guess compiling all of our code in the static code analyzers makes sense

File: test/teste2efiletransfer.cpp (hunk 6 above)
Comment (mgallien) [reply test/teste2efiletransfer.cpp:76]: do not we take the risk of random failures ?
how does the files client for android handle this ?

**Comment by claucambra** [comment]
> how would you run the tests locally on a dev setup ?

You could either:

- Set up drone locally
- Pull the docker container from our docker-ci images and point `endtoendtestutils.cpp` to the address which the local docker sets up

File: test/teste2efiletransfer.cpp (hunk 6 above)
Comment (claucambra) [reply test/teste2efiletransfer.cpp:76]: Android tests also make use of waiting through "sleep":

https://github.com/nextcloud/android/blob/90306642627e878789570a96bc415f0dd18f5d2e/app/src/androidTest/java/com/owncloud/android/FileIT.java#L117

https://github.com/nextcloud/android/blob/90306642627e878789570a96bc415f0dd18f5d2e/app/src/androidTest/java/com/owncloud/android/files/services/FileUploaderIT.kt#L151

File: .drone.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (claucambra) [reply .drone.yml:12]: fair enough, will change

File: .drone.yml [hunk 8]
```diff
@@ -73,6 +99,30 @@ steps:
...
+      - su www-data -c "OC_PASS=user3 php /var/www/html/occ user:add --password-from-env --display-name='User Three' user3"
+      - su www-data -c "php /var/www/html/occ user:setting user2 files quota 1G"
+      - su www-data -c "php /var/www/html/occ group:add users"
+      - su www-data -c "php /var/www/html/occ group:adduser users user1"
+      - su www-data -c "php /var/www/html/occ group:adduser users user2"
+      - su www-data -c "git clone -b $SERVER_VERSION https://github.com/nextcloud/activity.git /var/www/html/apps/activity/"
+      - su www-data -c "php /var/www/html/occ app:enable activity"
+      - su www-data -c "git clone -b $SERVER_VERSION https://github.com/nextcloud/text.git /var/www/html/apps/text/"
+      - su www-data -c "php /var/www/html/occ app:enable text"
+      - su www-data -c "git clone -b $SERVER_VERSION https://github.com/nextcloud/end_to_end_encryption.git /var/www/html/apps/end_to_end_encryption/"
```
[the remark this answers is not in the document]
Comment (claucambra) [reply .drone.yml]: Now --depth=1

File: .drone.yml [hunk 9]
```diff
@@ -29,6 +31,30 @@ steps:
...
```
[the remark this answers is not in the document]
Comment (claucambra) [reply .drone.yml]: Now --depth=1

File: test/teste2efiletransfer.cpp (hunk 2 above)
Comment (claucambra) [reply test/teste2efiletransfer.cpp:34]: now inits to a nullptr

File: test/endtoendtestutils.cpp (hunk 4 above)
Comment (claucambra) [reply test/endtoendtestutils.cpp]: Done

File: test/endtoendtestutils.cpp (hunk 1 above)
Comment (claucambra) [reply test/endtoendtestutils.cpp]: just a pointer now

Comment (claucambra) [reply test/endtoendtestutils.cpp:63]: Now we check for the validity

File: .drone.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (claucambra) [reply .drone.yml]: Done now

**Review by mgallien — APPROVED** [review]
